File: runtime/native/http_module.go
package native

import (
	"io"
	"net/http"
	"strings"

	nruntime "novago/runtime"
)

func CreateHTTPModule() *nruntime.ObjectVal {
	httpProps := make(map[string]nruntime.RuntimeVal)

	makeRequest := func(method string, urlStr string, bodyData string, headers map[string]string) nruntime.RuntimeVal {
		var reqBody io.Reader
		if bodyData != "" {
			reqBody = strings.NewReader(bodyData)
		}
		req, err := http.NewRequest(method, urlStr, reqBody)
		if err != nil {
			return nruntime.MK_NULL()
		}

		for k, v := range headers {
			req.Header.Set(k, v)
		}

		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			return nruntime.MK_NULL()
		}
		defer resp.Body.Close()

		resBody, _ := io.ReadAll(resp.Body)

		resProps := make(map[string]nruntime.RuntimeVal)
		resProps["data"] = nruntime.MK_STRING(string(resBody))
		resProps["status"] = nruntime.MK_NUMBER(float64(resp.StatusCode))

		return nruntime.MK_OBJECT(resProps)
	}

	httpProps["get"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		url := args[0].(*nruntime.StringVal).Value
		headers := make(map[string]string)
		if len(args) > 1 {
			if obj, ok := args[1].(*nruntime.ObjectVal); ok {
				for k, v := range obj.Properties {
					if strVal, ok := v.(*nruntime.StringVal); ok {
						headers[k] = strVal.Value
					}
				}
			}
		}
		return makeRequest("GET", url, "", headers)
	})

	httpProps["post"] = nruntime.MK_NATIVE_FN(func(args []nruntime.RuntimeVal, env *nruntime.Environment) nruntime.RuntimeVal {
		url := args[0].(*nruntime.StringVal).Value
		data := ""
		if len(args) > 1 {
			data = args[1].(*nruntime.StringVal).Value
		}
		headers := map[string]string{"Content-Type": "application/json"}
		if len(args) > 2 {
			if obj, ok := args[2].(*nruntime.ObjectVal); ok {
				for k, v := range obj.Properties {
					if strVal, ok := v.(*nruntime.StringVal); ok {
						headers[k] = strVal.Value
					}
				}
			}
		}
		return makeRequest("POST", url, data, headers)
	})

	return nruntime.MK_OBJECT(httpProps)
}

File: novascript/src/runtime/native/http_module.ts

import { RuntimeVal, MK_NATIVE_FN, MK_STRING, MK_BOOL, MK_NULL, MK_NUMBER, MK_OBJECT, FunctionVal, ReturnException } from "../values";
import { evaluate } from "../interpreter";
import Environment from "../environment";
import { execSync } from "child_process";
import * as http from "http";
import * as https from "https";
import { URL } from "url";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

// A more robust, native HTTP implementation for v5.0.0
export function createHTTPModule() {
    const httpProps = new Map<string, RuntimeVal>();

    // Native Client
    const makeRequest = (method: string, urlStr: string, data?: string, headers: Record<string, string> = {}) => {
        const script = `
            const { request } = require('${urlStr.startsWith('https') ? 'https' : 'http'}');
            const data = ${data ? JSON.stringify(data) : 'null'};
            const options = {
                method: '${method.toUpperCase()}',
                headers: ${JSON.stringify(headers)}
            };
            const req = request('${urlStr}', options, (res) => {
                let body = '';
                res.setEncoding('utf8');
                res.on('data', (chunk) => body += chunk);
                res.on('end', () => {
                    console.log(JSON.stringify({
                        data: body,
                        status: res.statusCode,
                        headers: res.headers
                    }));
                });
            });
            req.on('error', (e) => {
                console.log(JSON.stringify({ error: e.message }));
            });
            if (data) {
                req.write(data);
            }
            req.end();
        `;
        
        try {
            const tempFile = path.join(os.tmpdir(), `novascript-http-${Date.now()}.js`);
            fs.writeFileSync(tempFile, script);
            
            const nodeModulesPath = path.resolve(__dirname, '..', '..', '..', 'node_modules');
            const outputBuffer = execSync(`node "${tempFile}"`, {
                env: { ...process.env, NODE_PATH: nodeModulesPath }
            });
            const output = outputBuffer ? outputBuffer.toString() : "";
            fs.unlinkSync(tempFile);
            
            if (!output) {
                console.log("HTTP Request Error: No output from helper script.");
                return MK_NULL();
            }

            const res = JSON.parse(output);

            if (res.error) {
                console.log("HTTP Request Error:", res.error);
                return MK_NULL();
            }

            const resProps = new Map<string, RuntimeVal>();
            resProps.set("data", MK_STRING(res.data));
            resProps.set("status", MK_NUMBER(res.status));
            // You could add headers here too if needed
            return MK_OBJECT(resProps);
        } catch (e: any) {
            console.log("Failed to execute HTTP request:", e.message);
            return MK_NULL();
        }
    };

    httpProps.set("get", MK_NATIVE_FN((args) => {
        const url = (args[0] as any).value;
        const headersVal = args[1] && args[1].type === "object" ? (args[1] as any).properties : undefined;
        const headers: Record<string, string> = {};
        if (headersVal) {
            headersVal.forEach((v: any, k: string) => headers[k] = v.value);
        }
        return makeRequest("GET", url, undefined, headers);
    }));

    httpProps.set("post", MK_NATIVE_FN((args) => {
        const url = (args[0] as any).value;
        const data = (args[1] as any).value;
        const headersVal = args[2] && args[2].type === "object" ? (args[2] as any).properties : undefined;
        const headers: Record<string, string> = { 'Content-Type': 'application/json' };
        if (headersVal) {
            headersVal.forEach((v: any, k: string) => headers[k] = v.value);
        }
        return makeRequest("POST", url, data, headers);
    }));

    // Native Server
    httpProps.set("createServer", MK_NATIVE_FN((args, env) => {
        const routes = new Map<string, Map<string, FunctionVal>>();

        const server = http.createServer((req, res) => {
            const parsedUrl = new URL(req.url || "/", `http://${req.headers.host}`);
            const path = parsedUrl.pathname;
            const method = (req.method || "GET").toUpperCase();
            
            const pathRoutes = routes.get(path);
            const handler = pathRoutes ? pathRoutes.get(method) : undefined;

            if (!handler) {
                res.writeHead(404, { "Content-Type": "text/plain" });
                res.end("Not Found");
                return;
            }

            let body = "";
            req.on('data', chunk => body += chunk.toString());
            req.on('end', () => {
                const scope = new Environment(handler.declarationEnv);
                
                const reqProps = new Map<string, RuntimeVal>();
                reqProps.set("body", MK_STRING(body));
                // Add query/params later if needed
                const nsReq = MK_OBJECT(reqProps);
                
                const resProps = new Map<string, RuntimeVal>();
                resProps.set("send", MK_NATIVE_FN((args) => {
                    const data = (args[0] as any).value;
                    const headersVal = args[1] && args[1].type === "object" ? (args[1] as any).properties : undefined;
                    
                    if (!res.headersSent) {
                        const h: Record<string, string> = { "Content-Type": "text/plain" };
                        if (headersVal) {
                            headersVal.forEach((v: any, k: string) => {
                                h[k] = v.value;
                            });
                        }
                        res.writeHead(200, h);
                    }
                    res.end(data);
                    return MK_NULL();
                }));
                const nsRes = MK_OBJECT(resProps);

                if (handler.parameters.length >= 1) scope.declareVar(handler.parameters[0], nsReq, false);
                if (handler.parameters.length >= 2) scope.declareVar(handler.parameters[1], nsRes, false);

                try {
                    evaluate({ kind: "Program", body: handler.body } as any, scope);
                } catch (e) {
                    if (!(e instanceof ReturnException)) {
                        console.error("Error in NovaScript HTTP handler:", e);
                        res.writeHead(500);
                        res.end("Internal Server Error");
                    }
                }
            });
        });

        const serverObj = new Map<string, RuntimeVal>();
        const addRoute = (method: string) => MK_NATIVE_FN((args) => {
            const path = (args[0] as any).value;
            const handler = args[1] as FunctionVal;
            if (!routes.has(path)) routes.set(path, new Map());
            routes.get(path)!.set(method, handler);
            return MK_NULL();
        });

        serverObj.set("get", addRoute("GET"));
        serverObj.set("post", addRoute("POST"));

        serverObj.set("listen", MK_NATIVE_FN((args) => {
            const port = (args[0] as any).value;
            server.listen(port, () => {
                console.log(`NovaScript Native Server listening on port ${port}`);
            });
            // Keep process alive
            return MK_NULL();
        }));

        return MK_OBJECT(serverObj);
    }));

    return MK_OBJECT(httpProps);
}
